free-tunes/LogIn/LogInViewModel.swift

```diff
@@ -0,0 +1,54 @@
+//
+//  LogInViewModel.swift
+//  Free-tunes
+//
+//  Created by Tshwarelo Mafaralala on 2022/03/24.
+//
+
+import Foundation
+
+protocol LogInViewModelDelegate: AnyObject {
+    func notifyEmptyData()
+    func navigateToNext()
+    func show(error: String)
+}
+
+class LogInViewModel {
+    
+    private weak var delegate: LogInViewModelDelegate?
+    
+    init(delegate: LogInViewModelDelegate) {
+        self.delegate = delegate
+    }
+    
+// MARK: - Authentication
+    func login(username: String, password: String) {
+        if username.isEmpty || password.isEmpty {
+            delegate?.notifyEmptyData()
+            return
+        }
+        
+        if checkTextFieldData(usernameInput: username, passwordInput: password) {
+            if verifyCredentials(username: username, password: password) {
+                delegate?.navigateToNext()
+            }
+        }
+    }
+
+    private func verifyCredentials(username: String, password: String) -> Bool {
+        if username != "Admin" || password != "TestPass123" {
```

Comment:
Very nice `ViewModel` 😄 one suggestion that might make things a little bit easier, why not break down these functions a little bit more -they are sort of doing too much. 

Why don't you build up some `private` `vars` each of them with individual checks:

for example: 
```
private var validCredentials: Bool { 
   return username != "Admin" || password != "TestPass123"
}
```

And then you can have one `func` where you decide what `delegate` `func`  you need to call 😄 let me know if you need more clarification.

Reply:
since the username and password are passed in parameters would it make sense to make properties for them?

Reply:
So you'll need to keep one function where you pass them in and then use the `private` properties to do some checks in that function 😄

Reply:
ok i think I grasp the concept now😄